File: src/domain/providerUsage.ts
import { z } from "zod"

export function publicProviderUsageRouteLabel(raw: string): string {
  if (raw === "parser_time" || raw.startsWith("parser_time:")) {
    return "parser_time"
  }
  if (raw === "provider_candidate" || raw === "provider_route" || raw === "candidate_route") {
    return "provider_candidate"
  }
  if (
    raw === "provider_rejection" ||
    raw === "provider_rejected" ||
    raw.startsWith("provider_rejected:") ||
    raw === "provider_unavailable"
  ) {
    return "provider_rejection"
  }
  if (
    raw === "parser_pattern" ||
    raw === "calendar_route" ||
    raw === "quiet_route" ||
    raw === "no_action" ||
    raw.startsWith("parser_provider_route_") ||
    raw.startsWith("parser_pattern:")
  ) {
    return "parser_pattern"
  }
  return "unknown"
}

export const providerUsageWindowKeySchema = z.union([
  z.literal("7d"),
  z.literal("30d"),
  z.literal("90d"),
  z.literal("all")
])

const optionalNativeNumberSchema = z.preprocess(
  (value) => (value === null ? undefined : value),
  z.number().optional()
)

const optionalNativeConfidenceSchema = z.preprocess(
  (value) => (value === null ? undefined : value),
  z.number().min(0).max(1).optional()
)

const providerUsageGroupBaseSchema = z
  .object({
    totalOutcomes: z.number().int().min(0),
    candidateCount: z.number().int().min(0),
    quietCount: z.number().int().min(0),
    quietRate: z.number().min(0).max(1),
    candidateRate: z.number().min(0).max(1),
    averageConfidence: optionalNativeConfidenceSchema
  })
  .strict()

const providerUsagePromptGroupSchema = providerUsageGroupBaseSchema
  .extend({
    promptVersion: z.string().min(1)
  })
  .strict()

const providerUsageModelGroupSchema = providerUsageGroupBaseSchema
  .extend({
    modelId: z.string().min(1),
    promptVersions: z.array(providerUsagePromptGroupSchema)
  })
  .strict()

const providerUsageProviderGroupSchema = providerUsageGroupBaseSchema
  .extend({
    providerId: z.string().min(1),
    models: z.array(providerUsageModelGroupSchema)
  })
  .strict()

const providerUsageRecentOutcomeSchema = z
  .object({
    providerId: z.string().min(1),
    modelId: z.string().min(1),
    promptVersion: z.string().min(1),
    outcomeKind: z.string().min(1),
    routeLabel: z.string().min(1).transform(publicProviderUsageRouteLabel),
    confidence: optionalNativeConfidenceSchema,
    createdAtUnixSeconds: z.number().int().min(0)
  })
  .strict()

const providerUsageWindowSchema = z
  .object({
    key: providerUsageWindowKeySchema,
    startUnixSeconds: optionalNativeNumberSchema,
    endUnixSeconds: z.number().int().min(0)
  })
  .strict()

const providerUsageTotalsSchema = providerUsageGroupBaseSchema.omit({
  averageConfidence: true
})

export const providerUsageReportSchema = z
  .object({
    generatedAtUnixSeconds: z.number().int().min(0),
    window: providerUsageWindowSchema,
    totals: providerUsageTotalsSchema,
    providers: z.array(providerUsageProviderGroupSchema),
    recentOutcomes: z.array(providerUsageRecentOutcomeSchema).max(25)
  })
  .strict()

export const providerUsageLoadRequestSchema = z
  .object({
    windowKey: providerUsageWindowKeySchema.optional()
  })
  .strict()

export type ProviderUsageWindowKey = z.infer<typeof providerUsageWindowKeySchema>
export type ProviderUsageLoadRequest = {
  readonly windowKey?: ProviderUsageWindowKey | undefined
}

export type ProviderUsageTotals = {
  readonly totalOutcomes: number
  readonly candidateCount: number
  readonly quietCount: number
  readonly quietRate: number
  readonly candidateRate: number
}

export type ProviderUsagePromptGroup = ProviderUsageTotals & {
  readonly promptVersion: string
  readonly averageConfidence?: number | undefined
}

export type ProviderUsageModelGroup = ProviderUsageTotals & {
  readonly modelId: string
  readonly promptVersions: readonly ProviderUsagePromptGroup[]
  readonly averageConfidence?: number | undefined
}

export type ProviderUsageProviderGroup = ProviderUsageTotals & {
  readonly providerId: string
  readonly models: readonly ProviderUsageModelGroup[]
  readonly averageConfidence?: number | undefined
}

export type ProviderUsageWindow = {
  readonly key: ProviderUsageWindowKey
  readonly startUnixSeconds?: number | undefined
  readonly endUnixSeconds: number
}

export type ProviderUsageRecentOutcome = {
  readonly providerId: string
  readonly modelId: string
  readonly promptVersion: string
  readonly outcomeKind: string
  readonly routeLabel: string
  readonly confidence?: number | undefined
  readonly createdAtUnixSeconds: number
}

export type ProviderUsageReport = {
  readonly generatedAtUnixSeconds: number
  readonly window: ProviderUsageWindow
  readonly totals: ProviderUsageTotals
  readonly providers: readonly ProviderUsageProviderGroup[]
  readonly recentOutcomes: readonly ProviderUsageRecentOutcome[]
}

export function parseProviderUsageReport(value: unknown): ProviderUsageReport {
  return providerUsageReportSchema.parse(value)
}

export function parseProviderUsageLoadRequest(value: unknown): ProviderUsageLoadRequest {
  return providerUsageLoadRequestSchema.parse(value ?? {})
}

File: crates/morrow-storage/src/provider_usage.rs
pub const PROVIDER_USAGE_RECENT_OUTCOME_LIMIT: usize = 25;

const DAY_SECONDS: i64 = 24 * 60 * 60;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum ProviderUsageWindowKey {
    SevenDays,
    ThirtyDays,
    NinetyDays,
    All,
}

impl ProviderUsageWindowKey {
    pub const fn as_str(self) -> &'static str {
        match self {
            Self::SevenDays => "7d",
            Self::ThirtyDays => "30d",
            Self::NinetyDays => "90d",
            Self::All => "all",
        }
    }

    pub(crate) const fn start_unix_seconds(self, now_unix_seconds: i64) -> Option<i64> {
        match self {
            Self::SevenDays => Some(now_unix_seconds - (7 * DAY_SECONDS)),
            Self::ThirtyDays => Some(now_unix_seconds - (30 * DAY_SECONDS)),
            Self::NinetyDays => Some(now_unix_seconds - (90 * DAY_SECONDS)),
            Self::All => None,
        }
    }
}

#[derive(Debug, Clone, PartialEq)]
pub struct ProviderUsageReport {
    pub generated_at_unix_seconds: i64,
    pub window: ProviderUsageWindow,
    pub totals: ProviderUsageTotals,
    pub providers: Vec<ProviderUsageProvider>,
    pub recent_outcomes: Vec<ProviderUsageRecentOutcome>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ProviderUsageWindow {
    pub key: ProviderUsageWindowKey,
    pub start_unix_seconds: Option<i64>,
    pub end_unix_seconds: i64,
}

#[derive(Debug, Clone, PartialEq)]
pub struct ProviderUsageTotals {
    pub total_outcomes: u32,
    pub candidate_count: u32,
    pub quiet_count: u32,
    pub quiet_rate: f64,
    pub candidate_rate: f64,
}

#[derive(Debug, Clone, PartialEq)]
pub struct ProviderUsageProvider {
    pub provider_id: String,
    pub total_outcomes: u32,
    pub candidate_count: u32,
    pub quiet_count: u32,
    pub quiet_rate: f64,
    pub candidate_rate: f64,
    pub average_confidence: Option<f64>,
    pub models: Vec<ProviderUsageModel>,
}

#[derive(Debug, Clone, PartialEq)]
pub struct ProviderUsageModel {
    pub model_id: String,
    pub total_outcomes: u32,
    pub candidate_count: u32,
    pub quiet_count: u32,
    pub quiet_rate: f64,
    pub candidate_rate: f64,
    pub average_confidence: Option<f64>,
    pub prompt_versions: Vec<ProviderUsagePromptVersion>,
}

#[derive(Debug, Clone, PartialEq)]
pub struct ProviderUsagePromptVersion {
    pub prompt_version: String,
    pub total_outcomes: u32,
    pub candidate_count: u32,
    pub quiet_count: u32,
    pub quiet_rate: f64,
    pub candidate_rate: f64,
    pub average_confidence: Option<f64>,
}

#[derive(Debug, Clone, PartialEq)]
pub struct ProviderUsageRecentOutcome {
    pub provider_id: String,
    pub model_id: String,
    pub prompt_version: String,
    pub outcome_kind: String,
    pub route_label: String,
    pub confidence: Option<f64>,
    pub created_at_unix_seconds: i64,
}
